### ISSUE TITLE
New-M365DSCDeltaReport - No discrepancies have been found

### ISSUE THREAD

**Issue by adambrubaker** [microsoft/Microsoft365DSC#4948]
### Description of the issue

I'm doing some testing and manually generating changes to validate this tool will create alerts for our team when changes are made.  None of these 3 changes are showing up when I run a New-M365DSCDeltaReport.  I'm able to see the changes in the files by doing a compare in Notepad++.
1. Updating an existing EXOTransportRule
2. Creating a new EXOTransportRule
3. Updating an attribute in ODSettings

I just updated to the newest version and ran Update-M365DSCDependencies and Update-M365DSCModule.

### Microsoft 365 DSC Version

1.24.731.1

### Which workloads are affected

other

### The DSC configuration

```powershell
New-M365DSCDeltaReport -Source 'C:\temp\dsc\dsc_config.06182024.ps1' -Destination 'C:\temp\dsc\dsc_config.06192024.ps1' -OutputPath 'C:\temp\dsc\deltareport.html'

I don't mind uploading my config files, but I don't really want to attach them to a public case.  Is there a good way to upload them and not have them available to anyone who is following along?
```

### Verbose logs showing the problem

```powershell
VERBOSE: Obtaining Delta between the source and destination configurations
VERBOSE: Loading file 'C:\temp\dsc\dsc_config.06182024.ps1'
VERBOSE: Loading file 'C:\temp\dsc\dsc_config.06192024.ps1'
```

### Environment Information + PowerShell Version

```powershell
OsName               : Microsoft Windows 11 Enterprise
OsOperatingSystemSKU : EnterpriseEdition
OsArchitecture       : 64-bit
WindowsVersion       : 2009
WindowsBuildLabEx    : 22621.1.amd64fre.ni_release.220506-1250
OsLanguage           : en-US
OsMuiLanguages       : {en-US}

Key   : PSVersion
Value : 5.1.22621.3880
Name  : PSVersion

Key   : PSEdition
Value : Desktop
Name  : PSEdition

Key   : PSCompatibleVersions
Value : {1.0, 2.0, 3.0, 4.0...}
Name  : PSCompatibleVersions

Key   : BuildVersion
Value : 10.0.22621.3880
Name  : BuildVersion

Key   : CLRVersion
Value : 4.0.30319.42000
Name  : CLRVersion

Key   : WSManStackVersion
Value : 3.0
Name  : WSManStackVersion

Key   : PSRemotingProtocolVersion
Value : 2.3
Name  : PSRemotingProtocolVersion

Key   : SerializationVersion
Value : 1.1.0.1
Name  : SerializationVersion
```

**Comment by FabienTschanz** [comment]
@adambrubaker It would be very helpful to have the configuration. What you can do is anonymize and remove some of the configuration but still have the error be reproducible. Just a minimal amount of configuration so that we can have a look at that, doesn't have to be the complete one.

**Comment by adambrubaker** [comment]
OK so strange stuff here.  I deleted all sections that matched and just left the 3 that were different.  I kept all the header and footer information, open/close brackets, etc.

When I run the New-M365DSCDeltaReport command against these new files with only the 3 sections it produces a perfect report that shows all 3 differences.  

So I'm only able to reproduce the results with a bunch of other config info that is the same.  Does that make any sense?  I can probably anonymize the full configs, but it will take me a while to get them both updated.  Any other ideas?

**Comment by FabienTschanz** [comment]
@adambrubaker Sounds very interesting tbh, but for a full investigation we unfortunately need the configuration from you. It would be really helpful if you could provide that. I think removing anything that could identify the company you're working for should already be enough for the anonymization part, since the ids and the selectable configuration cannot be traced back to you.

**Comment by adambrubaker** [comment]
Totally understand.  I will get the files updated, make sure they still throw the error, and send them over.  Thanks!

**Comment by adambrubaker** [comment]
OK I got them cleaned up and they still show "No discrepancies have been found".  I had to update the file extensions to .txt so they would upload.  They were originally .ps1 files.  Thanks!


[dsc_config.06182024.txt](https://github.com/user-attachments/files/16502031/dsc_config.06182024.txt)
[dsc_config.06192024.txt](https://github.com/user-attachments/files/16502032/dsc_config.06192024.txt)

**Comment by FabienTschanz** [comment]
@adambrubaker Thanks a lot for the configuration. I already found an issue in the resource for `EXOTransportRule "EXOTransportRule-Mail_Delivery"`, there the property `SubjectOrBodyContainsWords` (which is an array) has an invalid escaped string: `""Remote server returned '554 5.7.0 < #5.7.520 smtp;","Your organization does not allow external forwarding. Please contact your administrator for further assistance"`. At the beginning of that string, there is a single `"` missing. 

If I add that and retry the comparison, three differences are outputted: 

- EXOTransportRule-Message Tagging for Company Tenants
- ODSettings
- EXOTransportRule-Adam Testing

Can you try and do the same on your side, just to check if it also works?

Edit: Did you export the configuration that way? Then maybe there is an issue with the DSCParser that we would have to check.

**Comment by adambrubaker** [comment]
@FabienTschanz thanks for the quick response.  That was the issue!  After removing those quotes I'm able to see the changes in my configs using New-M365DSCDeltaReport now.  I'm glad it was an easy fix.  To answer your question the quotes were in our config.  It wasn't an issue with the export.

Question.  If I see something similar again is there any way I can troubleshoot this on my own?  Are you leveraging some other module I can download and use or is it something custom in the Microsoft365DSC module?

For those playing along at home the quotes in front of the word _Remote_ in my Exchange Transport Rule was causing the issue.  I just removed that first character.  Screenshot below.

![image](https://github.com/user-attachments/assets/c65ab8b1-1c4b-4162-b79a-e55c36a98adb)

**Comment by FabienTschanz** [comment]
@adambrubaker Glad we got it worked out. The basic troubleshooting can be done by using VS Code and the PowerShell extension, which highlights potential issues with your configuration since there is syntax highlighting and an automatic check of the configuration when it is opened in the editor. For additional information during the runtime (e.g. when using `Export-M365DSCConfiguration` or something like that), you can add the `-Verbose` switch parameter to the command and you might get more information from that output. 

But it's always quite good to open an issue, since there might be another problem in place than just something "obvious" (like the double quotes in your case here). So feel free to drop by and ask 😄

**Comment by adambrubaker** [comment]
Ahh... that is a good first step.  Easy to see in VS Code.  Thanks again!

**Comment by chemical1979** [comment]
I was also doing some tests with this great documentation solution, and run into the same situation. I'm just using the AADConditionalAccessPolicy. I did an export, created an additional policy, and afterwards exported again (same versions like adambrubaker). 

I did a quick check for missing quotes or missing escapes, but did not find any. I did the export with `-Verbose` but did not notice anything strange - same with the Delta on verbose. 

I get an error with one of the policies that Microsoft enforced into every tenant some time ago, although it is turned off.
`{InvalidOperation}
System.Exception: [Request_ResourceNotFound] : Resource '1874a4fb-7cdc-41f0-aca6-cc11e8b42b63' does not exist or one of its queried reference-property objects are not present.
"Couldn't find IncludedUser '1874a4fb-7cdc-41f0-aca6-cc11e8b42b63', that is defined in policy 'Microsoft-managed: Multifactor authentication for per-user multifactor authentication users'. Skipping user."
at Get-MgUser<Process>, C:\Program Files\WindowsPowerShell\Modules\Microsoft.Graph.Users\2.20.0\exports\ProxyCmdletDefinitions.ps1: line 18602
at Get-TargetResource, C:\Program Files\WindowsPowerShell\Modules\Microsoft365DSC\1.24.731.1\DSCResources\MSFT_AADConditionalAccessPolicy\MSFT_AADConditionalAccessPolicy.psm1: line 307
at Export-TargetResource, C:\Program Files\WindowsPowerShell\Modules\Microsoft365DSC\1.24.731.1\DSCResources\MSFT_AADConditionalAccessPolicy\MSFT_AADConditionalAccessPolicy.psm1: line 2071
at Start-M365DSCConfigurationExtract, C:\Program Files\WindowsPowerShell\Modules\Microsoft365DSC\1.24.731.1\Modules\M365DSCReverse.psm1: line 682
at Export-M365DSCConfiguration, C:\Program Files\WindowsPowerShell\Modules\Microsoft365DSC\1.24.731.1\Modules\M365DSCUtil.psm1: line 1394
at <ScriptBlock>, Export-ConditionalAccessPolicies.ps1: line 16
at <ScriptBlock>, <No file>: line 1
TenantId: XXX.onmicrosoft.com`
Sadly, this policy is protected and somehow internally damaged (it is showing 3 users in scope, but only 1 of I want to change it. I am also not possible to Save it, as it tells me I am not allowed to modify the policy except for excluding users).

I tried to debug it in VS Code, placed a few breakpoints and I was under the impression the Delta compares two Null arrays, which means it fails to import the structures of both files somehow.

Any ideas on how to narrow it down or where to put the breakpoint exactly? I got lost in the Stack somewhere :-) 
The only thing that puzzled me was that in `ConvertTo-DSCObject` the `$Content` was still fine, and after the Parsing the Configuration definition `$Config` was empty:

`$AST = [System.Management.Automation.Language.Parser]::ParseInput($Content, [ref]$Tokens, [ref]$ParseErrors)`
`# Look up the Configuration definition ("")`
`$Config = $AST.Find({$Args[0].GetType().Name -eq 'ConfigurationDefinitionAst'}, $False)`

Subsequently, no Modules were loaded, and all other following steps rely on `$Config`.
